File: Leetcode/src/com/leetcode/_42_TrappingRainWater.java
package com.leetcode;

public class _42_TrappingRainWater {

    /**
     * 42. Trapping Rain Water
     * when: 2019/03/25
     *
     * solution: two-pointer
     * 分别设置左右max，然后与当前的进行相差则就是对应的乘水量
     * @param height
     * @return
     */
    public static int trap(int[] height) {
        int res = 0;
        int left = 0;
        int right = height.length - 1;
        int leftMax = 0;
        int rightMax = 0;
        while (left < right){
            if (height[left] < height[right]){
                leftMax = Math.max(leftMax, height[left]);
                res += leftMax - height[left];
                left++;
            }
            else {
                rightMax = Math.max(rightMax, height[right]);
                res += rightMax - height[right];
                right--;
            }
        }
        return res;
    }

    public static void main(String[] args){
        int[] height = new int[]{0,1,0,2,1,0,1,3,2,1,2,1};
        System.out.println(trap(height));
    }
}

File: Leetcode/src/com/leetcode/array/twoPointer/_42_trapping_rain_water.py
from typing import List


"""
42. Trapping Rain Water
---

Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
---

Example 1:
Input: height = [0,1,0,2,1,0,1,3,2,1,2,1]
Output: 6
Explanation: The above elevation map (black section) is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped.

Example 2:
Input: height = [4,2,0,3,2,5]
Output: 9
"""


class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        left, right = 0, n - 1
        left_max, right_max = 0, 0
        res = 0
        while left < right:
            left_h, right_h = height[left], height[right]
            if left_h <= right_h:
                left_max = max(left_max, height[left])
                res += left_max - height[left]
                left += 1
            else:
                right_max = max(right_max, height[right])
                res += right_max - height[right]
                right -= 1
        return res
